### PR TITLE
Model tree diagram

### PR DESCRIPTION
**Pull request by barrettk**
Introduces new `model_tree()` function, allowing you to create a tree diagram for your modeling directory

### Some Example function calls
```r
model_tree(model_dir)
model_tree(log_df)
model_tree(log_df, add_summary = FALSE)
model_tree(log_df, color_by = "star")
model_tree(log_df, static = TRUE)

log_df %>% add_summary() %>%
  model_tree(
    include_info = c("tags", "param_count", "eta_pval_significant"),
    color_by = "any_heuristics"
  )

log_df %>% add_config() %>%
  model_tree(
    include_info = c("model_has_changed", "data_has_changed", "nm_version"),
    color_by = "data_has_changed"
  )
```

### Example output
```r
model_tree(log_df,  color_by = "run")
```
image

Logical values will be colored as white and red, and are referred to as 'border colors'. The more unique values are added, the more gradient colors between those two border colors are generated.
```r
model_tree(log_df, color_by = "star")
```
image

This example illustrates how `NA` values are handled. Here we color by description (which wouldn't make sense to do in practice). The two distinct descriptions are colored white and red, where `NA` values are colored grey.
```r
run_log(model_dir) %>%
  add_config() %>%
  model_tree(
    include_info =  c("tags", "description", "model_has_changed", "data_has_changed"),
    color_by = "description", add_summary = FALSE
  )
```
image

Previous PR for reference: https://github.com/metrumresearchgroup/bbr/pull/610

### CODE REVIEW

**Comment by andersone1** [comment]
@barrettk 

Can `webshot::install_phantomjs()` be added to the top of `test-model-tree.R`

**Comment by andersone1** [comment]
- [ ] model_tree
- [ ] model_tree.character
- [ ] model_tree.bbi_log_df
- [ ] make_tree_data
- [ ] unnest_based_on
- [ ] make_model_network
- [ ] add_log_recurse_dirs
- [ ] check_model_tree
- [ ] make_tree_tooltip
- [ ] color_tree_by
- [ ] style_html
- [ ] format_model_type
- [ ] model_tree_png
- [ ] print.model_tree_static
- [ ] req_tree_pkgs
- [ ] check_for_model_tree_pkgs
- [ ] skip_if_tree_missing_deps
- [ ] stop_if_tree_missing_deps

**Review by andersone1 — APPROVED** [review]

File: R/model-tree.R [hunk 1]
```diff
@@ -0,0 +1,885 @@
...
+    # i.e. model_tree(run_log(model_dir) %>% add_summary(), include_info = c("problem_text", "ofv"))
+    #  - Remove any specially handled columns prior to running
+    #  - Only join the _new_ select summary columns back
+    sum_log <- full_log %>% dplyr::select(-any_of(sum_cols)) %>% add_summary() %>%
+      dplyr::select(all_of(c(req_cols, sum_cols)))
+    if(any(sum_cols %in% names(full_log))){
+      existing_sum_cols <- sum_cols[sum_cols %in% names(full_log)]
+      sum_log <- sum_log %>% dplyr::select(-all_of(existing_sum_cols))
+    }
+    full_log <- dplyr::left_join(full_log, sum_log, by = req_cols)
```
[the remark this answers is not in the document]
Comment (barrettk) [reply R/model-tree.R:290]: @andersone1 Shouldn't the join enforce that? If we join by `req_cols` I cant think of any way the number of rows would change...unless you made the `run_log` and then added a new model, before running `model_tree()`, though that seems like an edge case. Can you think of any other way that could happen?

**Comment by kyleam** [comment]
Most of the check failures are straightforward.  I can help take a look at the oldest job failure:

https://github.com/metrumresearchgroup/bbr/actions/runs/9701317535/job/26774698747

```
   ℹ Creating lockfile '.github/pkg.lock'
  
  ✔ Updated metadata database: 3.47 MB in 10 files.
  
  ℹ Creating lockfile '.github/pkg.lock'
  ℹ Updating metadata database
  ✔ Updating metadata database ... done
  
  ℹ Creating lockfile '.github/pkg.lock'
  ✖ Creating lockfile '.github/pkg.lock' [6.3s]
  
  Error: 
  ! error in pak subprocess
  Caused by error: 
  ! Could not solve package dependencies:
  * deps::.: Can't install dependency scales (>= 1.3.0)
  * scales: scales conflicts with scales, to be installed
  * any::sessioninfo: dependency conflict
  * any::pkgdown: dependency conflict
  * any::rcmdcheck: dependency conflict
```

**Comment by kyleam** [comment]
> scales (>= 1.3.0)

1.3.0 isn't on MPN 2023-01-25 (it wasn't released until Nov 2023).

@barrettk is this constraint just copied from ggplot2?  Or are we using scales directly in a way that requires 1.3.0 or later?  If the former, I think we can drop the constraint, and hopefully that will resolve the pak failure.

**Comment by barrettk** [comment]
@kyleam It wasn't copied from `ggplot2` and I believe I did that because of `pal_gradient_n`, and there wasn't really an easy workaround. That being said I do have functions set up to skip tests and not run the functions if that version isnt there, so dropping it may still be a good solution.

**Comment by kyleam** [comment]
> I believe I did that because of `scales::pal_gradient_n`

I see.  Then the options I can think of are

 * remove the constraint and add guards to check it at runtime for the specific spots that need a given version
 * install a newer scales from elsewhere (as we do for nmrec or mrgmisc)

I think you're going with the first option (no objections from me).

**Comment by kyleam** [comment]
Highlighting this note, since it won't fail the build:

https://github.com/metrumresearchgroup/bbr/actions/runs/9702122713/job/26777261543

``` 
* checking Rd \usage sections ... NOTE
Documented arguments not in \usage in Rd file 'format_model_type.Rd':
  ‘.mod’ ‘msg’
```

**Comment by barrettk** [comment]
@kyleam can you take a look at the CI before and after I added `collapsibleTree` to github actions? A) it seems like it was being installed on mpn-oldest before I added it, but B) clicking the green check is now displaying different/less options.

**Comment by kyleam** [comment]
> A) it seems like it was being installed on mpn-oldest before I added it

You don't need the custom step as long as it can be retrieved from the date-pinned RSPM URL.

**Comment by barrettk** [comment]
>You don't need the custom step as long as it can be retrieved from the date-pinned RSPM URL.

Would you recommend reverting this change then, or is there potentially a reason to keep it (some form of risk mitigation)?

**Comment by kyleam** [comment]
> Would you recommend reverting this change then

yes

File: R/model-tree.R (hunk 1 above)
[the remark this answers is not in the document]
Comment (andersone1) [reply R/model-tree.R:290]: @barrettk I can't really think of a way it could happen, but `left_join` by does not enforce that.

**Pull request merged**